Engine/Rendering/FrameBuffer/FrameBuffer.cpp

```diff
@@ -0,0 +1,126 @@
+#include "FrameBuffer.h"
+
+FrameBuffer::FrameBuffer(int num_color_attachements) : num_color_attachements(num_color_attachements)
+{
+	glGenFramebuffers(1, &fbo);
+}
+
+FrameBuffer::~FrameBuffer()
+{
+	ClearAttachements();
+	glDeleteFramebuffers(1, &fbo);
+}
+
+void FrameBuffer::GenerateAttachements(float width, float height)
+{
+	GenerateColorAttachement(width, height);
+	GenerateDepthAttachement(width, height);
+	LinkAttachements();
+}
+
+void FrameBuffer::ClearAttachements()
+{
+	for (size_t i = 0; i < num_color_attachements; ++i)
+	{
+		if (color_attachements[i] != 0)
+		{
+			glDeleteTextures(1, &color_attachements[i]);
+		}
+	}
+
+	if (rbo != 0)
+	{
+		glDeleteRenderbuffers(1, &rbo);
+	}
+}
+
+void FrameBuffer::Bind(GLenum target) const
+{
+	glBindFramebuffer(target, fbo);
+}
+
+void FrameBuffer::UnBind(GLenum target)
+{
+	glBindFramebuffer(target, 0);
+}
+
+GLuint FrameBuffer::GetColorAttachement(int index) const
+{
+	return color_attachements[index];
+}
+
+void FrameBuffer::ClearColorAttachement(GLenum color_attachement, const float3& clear_color, GLbitfield mask) const
+{
+	glDrawBuffer(color_attachement);
+	glClearColor(clear_color.x, clear_color.y, clear_color.z, 1.0f);
+	glStencilMask(0xFF);
+	glClear(mask);
+	glDrawBuffer(0);
+}
+
+void FrameBuffer::GenerateColorAttachement(float width, float height)
+{
+	glGenTextures(num_color_attachements, &color_attachements[0]);
+	for (size_t i = 0; i < num_color_attachements; ++i)
+	{
+		if (multisampled)
+		{
+			glBindTexture(GL_TEXTURE_2D_MULTISAMPLE, color_attachements[i]);
+			glTexImage2DMultisample(GL_TEXTURE_2D_MULTISAMPLE, 4, GL_RGBA16F, width, height, GL_TRUE);
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+			glBindTexture(GL_TEXTURE_2D_MULTISAMPLE, 0);
+		}
+		else
+		{
+			glBindTexture(GL_TEXTURE_2D, color_attachements[i]);
+			glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA16F, width, height, 0, GL_RGBA, GL_FLOAT, 0);
```

Comment:
you need 16 if you are going to use a high level of precision, in this case, hdr, but if you want a texture without that resolution, GL_RGBA will do the case , because you will use 8 bits instead of 16.